### PR TITLE
Add security patterns page

### PR DESCRIPTION
**Pull request by unknwon**
Creating this doc per request from @ElizabethStirling 

Every time I write a doc, I want your approval @quinnkeast 😁

### CODE REVIEW

File: handbook/engineering/security_patterns.md [hunk 1]
```diff
@@ -0,0 +1,23 @@
+# Security patterns
+
+## Authorization
+
+Authorization is the gatekeeper of securing private resources by preventing unauthorized access to those private resources. Examples of private resources are:
```
Comment (quinnkeast) [inline handbook/engineering/security_patterns.md:5]:
```suggestion
Authorization is the gatekeeper for securing private resources by preventing unauthorized access. Examples of private resources are: 
```

File: handbook/engineering/security_patterns.md [hunk 2]
```diff
@@ -0,0 +1,23 @@
+# Security patterns
+
+## Authorization
+
+Authorization is the gatekeeper of securing private resources by preventing unauthorized access to those private resources. Examples of private resources are: 
+
+- Content of a private repository
```
Comment (quinnkeast) [inline handbook/engineering/security_patterns.md:7]:
```suggestion
- Contents of a private repository
```

File: handbook/engineering/security_patterns.md [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+- Endpoints that can only be called by certain users (e.g. site admins)
+- Settings of a user or an organization
```
Comment (quinnkeast) [inline handbook/engineering/security_patterns.md:9]:
```suggestion
- User settings or organization settings
```

File: handbook/engineering/security_patterns.md [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+Current forms of authorization in Sourcegraph including site-admin roles, organization memberships, campaigns permissions, repository permissions, and same user validation.
```
Comment (quinnkeast) [inline handbook/engineering/security_patterns.md:11]:
```suggestion
Current forms of authorization in Sourcegraph include:

 - Site admin roles
 - Organization memberships
 - Campaigns permissions
 - Repository permissions
 - Same-user validation
```

File: handbook/engineering/security_patterns.md [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+As a standard practice, users that do not have access to certain private resources should not be aware of the existencce of those private resources. In the case where only part of resources are restricted by authorization, it is reasonable to prompt the user that some resources are not avaialble because of insufficent permissions. However, in any case, we must not provide anything that can be used to indicate what those restricted resources might be.
```
Comment (quinnkeast) [inline handbook/engineering/security_patterns.md:13]:
```suggestion
As a standard practice, users who do not have access to a given private resource should not be aware of the existence of that private resource. When only part of a resource is restricted by authorization limitations, it is reasonable to prompt the user that some resources are not available due to insufficient permissions. However, we must not provide anything that indicates what those restricted resources might be.
```

File: handbook/engineering/security_patterns.md [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+
+### Enforce authorization
+
+In Sourcegraph, there are two types of places to enforce authorization, both of them are equally important:
```
Comment (quinnkeast) [inline handbook/engineering/security_patterns.md:17]:
```suggestion
In Sourcegraph, there are two places to enforce authorization, both equally important:
```

File: handbook/engineering/security_patterns.md [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+
+- At the GraphQL layer:
+    - Some endpoints are restricted to certain users (e.g. site admins or the same user).
+    - Any type of backend failure may indicate information about private resources that are not authorized to the client. Therefore, we should halt the process as soon as we know the request is unauthorized, and behave as if the resource does not exist at all.
```
Comment (quinnkeast) [inline handbook/engineering/security_patterns.md:21]:
```suggestion
    - Be aware that any backend failure has potential to indicate unauthorized information about private resources. Therefore, halt the process as soon as we identify an unauthorized request, and behave as if the resource does not exist at all.
```

File: handbook/engineering/security_patterns.md [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+- At the database layer:
+    - It is our source of truth for authorization, especially for repository and campaigns permissions. Enforcing authorization at this layer is absolutely necessary.
```
Comment (quinnkeast) [inline handbook/engineering/security_patterns.md:23]:
```suggestion
    - The database our source of truth for authorization, especially for repository and campaigns permissions. Enforcing authorization at this layer is absolutely necessary.
```

You'll have to confirm this change is technically correct – is the source of truth the "database" or the "database layer?"

**Comment by unknwon** [comment]
Thanks @quinnkeast! Every time your suggestions would level up the writing of the doc :D

**Review by ElizabethStirling — APPROVED** [review]
Thanks so much for hammering this out!

**Pull request merged**